contrib/endpoints/src/api_manager/request_handler_test.cc

```diff
@@ -0,0 +1,619 @@
+  }
+
+  ::google::api_manager::protocol::Protocol GetFrontendProtocol() {
+    return ::google::api_manager::protocol::Protocol::HTTP;
+  }
+
+  ::google::api_manager::protocol::Protocol GetBackendProtocol() {
+    return ::google::api_manager::protocol::Protocol::HTTPS;
+  }
+
+  void SetAuthToken(const std::string &auth_token) {}
+
+  utils::Status AddHeaderToBackend(const std::string &key,
+                                   const std::string &value) {
+    return utils::Status::OK;
+  }
+
+ private:
+  std::unordered_map<std::string, std::string> data_;
+};
+
+class ResponseMock : public Response {
+ public:
+  virtual ~ResponseMock() {}
+  utils::Status GetResponseStatus() { return utils::Status::OK; }
+  std::size_t GetRequestSize() { return 0; }
+  std::size_t GetResponseSize() { return 0; }
+  utils::Status GetLatencyInfo(service_control::LatencyInfo *info) {
+    return utils::Status::OK;
+  }
+};
+
+class RequestHandlerTest : public ::testing::Test {
+ protected:
+  RequestHandlerTest() : callback_run_count_(0) {}
+  std::shared_ptr<ApiManager> MakeApiManager(
+      std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config);
+  std::shared_ptr<ApiManager> MakeApiManager(
+      std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config,
+      const char *server_config);
+
+  void SetUp() {
+    callback_run_count_ = 0;
+    call_history_.clear();
+  }
+
+ protected:
+  std::vector<std::string> call_history_;
+  int callback_run_count_;
+
+ private:
+  ApiManagerFactory factory_;
+};
+
+std::shared_ptr<ApiManager> RequestHandlerTest::MakeApiManager(
+    std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config) {
+  return factory_.CreateApiManager(std::move(env), service_config, "");
+}
+
+std::shared_ptr<ApiManager> RequestHandlerTest::MakeApiManager(
+    std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config,
+    const char *server_config) {
+  return factory_.CreateApiManager(std::move(env), service_config,
+                                   server_config);
+}
+
+TEST_F(RequestHandlerTest, PendingCheckApiManagerInitSucceeded) {
+  std::unique_ptr<MockApiManagerEnvironment> env(
+      new ::testing::NiceMock<MockApiManagerEnvironment>());
+
+  EXPECT_CALL(*(env.get()), DoRunHTTPRequest(_))
+      .WillRepeatedly(Invoke([this](HTTPRequest *req) {
+        std::unordered_map<std::string, std::string> data = {
+            {"https://servicemanagement.googleapis.com/v1/services/"
+             "bookstore.test.appspot.com/configs/2017-05-01r0",
+             kServiceConfig1},
+            {"http://localhost/computeMetadata/v1/?recursive=true", "{}"}};
+
+        call_history_.push_back(req->url());
+        if (data.find(req->url()) == data.end()) {
+          req->OnComplete(Status(Code::NOT_FOUND, "Not Found"), {},
+                          std::move(kServiceConfig1));
+        } else {
+          req->OnComplete(Status::OK, {}, std::move(kServiceConfig1));
```

Comment:
context->api_key() is empty and allowUnregisteredCalls is set to true
CheckServiceControl will not call the Chemist->Check()